lib/OpenQA/Parser/Format/KTAP.pm

```diff
@@ -0,0 +1,188 @@
+use Carp qw(croak confess);
+use OpenQA::Parser::Result::OpenQA;
+use TAP::Parser;
+
+has [qw(test steps)];
+
+sub _is_tap ($raw) {
+    my $tap = TAP::Parser->new({tap => $raw});
+    confess 'KTAP parse error: ' . $tap->parse_errors if $tap->parse_errors;
+
+    while (my $result = $tap->next) {
+        return $result->version if $result->type eq 'version';
+    }
+
+    return undef;
+}
+
+sub _testgroup_init ($self, $line, $test_ref, $steps_ref, $m_ref) {
+    if ($$steps_ref && $self->test) {
+        $$steps_ref->{name} = $self->test->{name};
+        $$steps_ref->{test} = $self->test;
+        $self->generated_tests_results->add(OpenQA::Parser::Result::OpenQA->new($$steps_ref));
+    }
+    my ($group_name) = $line =~ /^#\s*selftests:\s+(.*)/;
+    my $sanitized_group_name = "selftests: $group_name";
+    $sanitized_group_name =~ s/[\/.]/_/g;
+    $$test_ref = {
+        flags => {},
+        category => 'KTAP',
+        name => $sanitized_group_name,
+    };
+    $self->test(OpenQA::Parser::Result->new($$test_ref));
+    $self->_add_test($self->test);
+    $$steps_ref = OpenQA::Parser::Result->new(
+        {
+            details => [],
+            dents => 0,
+            result => 'passed'
+        });
+
+    $$m_ref = 0;
+}
+
+sub _parse_subtest ($self, $result, $test_ref, $steps_ref, $m_ref) {
+    return unless $$steps_ref && $self->test;
+    my $line = $result->as_string;
+    return unless $line =~ /^#\s*(ok|not ok)\s+\d+\s+(.+)/;
+    my ($status, $subtest_name) = ($1, $2);
+    my $filename = "KTAP-@{[$$test_ref->{name}]}-$$m_ref.txt";
+    push @{$$steps_ref->{details}},
+      {
+        text => $filename,
+        title => $subtest_name,
+        result => $status eq 'ok' ? 'ok' : 'fail',
+      };
+    $$steps_ref->{result} = 'fail' if $status ne 'ok';
+    $self->_add_output({file => $filename, content => $line});
+    $$m_ref++;
+}
+
+sub _testgroup_finalize ($self, $steps_ref, $result) {
+    return unless $$steps_ref && $self->test;
+
+    my $line = $result ? $result->as_string : '';
+    my $group_failed = 0;    # it is possible the subtests do not report 'not ok' but only the group
+    if ($line =~ /^not ok\b/i) {
+        $$steps_ref->{result} = 'fail';
+        $group_failed = 1;
+    }
+    elsif ($line =~ /#\s*SKIP\b/i) {
+        $$steps_ref->{result} = 'skip';
+    }
+    $$steps_ref->{name} = $self->test->{name};
+    $$steps_ref->{test} = $self->test;
+    $self->generated_tests_results->add(OpenQA::Parser::Result::OpenQA->new($$steps_ref));
+    $self->test(undef); # clear the group
+    $$steps_ref = undef; # clear the subtests
+}
+
+sub parse ($self, $KTAP) {
+    _is_tap($KTAP);
+
+    my $parser = TAP::Parser->new({tap => $KTAP});
+    my ($test, $steps, $m) = ({}, undef, 0);
+
+    while (my $result = $parser->next) {
+        next if $result->type eq 'version' || $result->type eq 'plan';
+        if ($result->type eq 'comment' && $result->as_string =~ /^#\s*selftests:\s+(.*)/) {
```

Comment:
there is a weird logic from my POV. 
Inside _is_tap you create the same object which here in L95 you recreate. it seems to me more intuitive to create the parser before _is_tap and pass the instance.

Reply:
Yes, it would probably make sense to use only one parser here. Or is iterating only possible once?

Reply:
v2 pushed.
I will resolve this for now